geode-cq/src/distributedTest/java/org/apache/geode/internal/cache/PutAllClientServerDistributedTest.java

```diff
@@ -0,0 +1,3344 @@
+import org.apache.geode.test.junit.categories.ClientSubscriptionTest;
+import org.apache.geode.test.junit.rules.serializable.SerializableTestName;
+
+/**
+ * Tests putAll for c/s. Also tests removeAll
+ *
+ * @since GemFire 5.0.23
+ */
+@Category({ClientServerTest.class, ClientSubscriptionTest.class})
+@RunWith(JUnitParamsRunner.class)
+@SuppressWarnings("serial")
+public class PutAllClientServerDistributedTest extends CacheTestCase {
+
+  private static final long TIMEOUT_MILLIS = GeodeAwaitility.getTimeout().getValueInMS();
+
+  private static final int ONE_HUNDRED = 100;
+  private static final int ONE_THOUSAND = 1000;
+
+  private static final CacheListenerAction<Integer> EMPTY_INTEGER_ACTION =
+      new CacheListenerAction<>(null, emptyConsumer());
+
+  private static final Counter DUMMY_COUNTER = new Counter("dummy");
+
+  private static volatile AtomicReference<TickerData> oldValueReference;
+  private static volatile CountDownLatch createdCqLatch;
+  private static volatile CountDownLatch validatedCqLatch;
+  private static volatile CountDownLatch operationLatch;
+  private static volatile Counter serverCounter;
+
+  private String regionName;
+  private String hostName;
+  private String title;
+  private String locators;
+
+  private VM server1;
+  private VM server2;
+  private VM client1;
+  private VM client2;
+
+  @Rule
+  public SerializableTestName testName = new SerializableTestName();
+
+  @Rule
+  public SharedErrorCollector errorCollector = new SharedErrorCollector();
+
+  @Before
+  public void setUp() {
+    regionName = getClass().getSimpleName();
+    hostName = getHostName();
+    title = testName.getMethodName();
+    locators = getLocators();
+
+    server1 = getVM(0);
+    server2 = getVM(1);
+    client1 = getVM(2);
+    client2 = getVM(3);
+
+    addIgnoredException(ConnectException.class);
+    addIgnoredException(PutAllPartialResultException.class);
+    addIgnoredException(RegionDestroyedException.class);
+    addIgnoredException(ServerConnectivityException.class);
+    addIgnoredException(SocketException.class);
+
+    addIgnoredException("Broken pipe");
+    addIgnoredException("Connection reset");
+    addIgnoredException("Unexpected IOException");
+  }
+
+  @After
+  public void tearDown() {
+    for (VM vm : toArray(client1, client2, server1, server2)) {
+      vm.invoke(() -> {
+        if (createdCqLatch != null) {
+          createdCqLatch.countDown();
+          createdCqLatch = null;
+        }
+        if (validatedCqLatch != null) {
+          validatedCqLatch.countDown();
+          validatedCqLatch = null;
+        }
+        if (operationLatch != null) {
+          operationLatch.countDown();
+          operationLatch = null;
+        }
+        disconnectFromDS();
+      });
+    }
+  }
+
+  /**
+   * Tests putAll to one server.
+   */
+  @Test
+  public void testOneServer() throws Exception {
+    // set <PR=false, notifyBySubscription=true> to enable registerInterest and CQ
```

Comment:
Can you make this test name more descriptive of what actually happens or add a comment?